Code/GraphMol/testSGroup.cpp

```diff
@@ -1018,13 +1021,180 @@ void testSubstanceGroupsAndRemoveHs(const std::string &rdbase) {
       ps.removeInSGroups = true;
       MolOps::removeHs(mol_copy, ps);
       TEST_ASSERT(mol_copy.getNumAtoms() == 6);
-      TEST_ASSERT(getSubstanceGroups(mol_copy).size() == 0);
+      TEST_ASSERT(getSubstanceGroups(mol_copy).size() == 1);
     }
     {  // check removeAllHs() too
       RWMol mol_copy = *mol;
       MolOps::removeAllHs(mol_copy);
       TEST_ASSERT(mol_copy.getNumAtoms() == 6);
-      TEST_ASSERT(getSubstanceGroups(mol_copy).size() == 0);
+      TEST_ASSERT(getSubstanceGroups(mol_copy).size() == 1);
+    }
+  }
+}
+
+void testKeepSpecialHsOnRemoval() {
```

Comment:
Just to make it explicit and clear, it's maybe worth adding a test where the H is outside the S group but at the end of an XBOND (I don't think you have this covered already):
```

  Mrv2102 07062206502D          

  0  0  0     0  0            999 V3000
M  V30 BEGIN CTAB
M  V30 COUNTS 4 3 1 0 0
M  V30 BEGIN ATOM
M  V30 1 H -10.875 -2.2083 0 0
M  V30 2 C -9.5413 -1.4383 0 0
M  V30 3 C -8.2076 -2.2083 0 0
M  V30 4 * -6.874 -1.4383 0 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 1 2
M  V30 2 1 2 3
M  V30 3 1 3 4
M  V30 END BOND
M  V30 BEGIN SGROUP
M  V30 1 SRU 0 ATOMS=(2 2 3) XBONDS=(2 1 3) BRKXYZ=(9 -8.1205 -1.0911 0 -
M  V30 -7.1965 -2.6915 0 0 0 0) BRKXYZ=(9 -9.6285 -2.5556 0 -10.5525 -0.9552 -
M  V30 0 0 0 0) CONNECT=HT LABEL=n
M  V30 END SGROUP
M  V30 END CTAB
M  END
```
I believe no atoms should be removed from this one.

Reply:
That is covered by the first case in the test: from the 4 H atoms in the `CH4`, 2 are part of the SGroup, and will get removed. The other two are in XBONDS, and will stay.